### DISCUSSION TITLE
Quterofi

### DISCUSSION THREAD

**Discussion by cortsf** [qutebrowser/qutebrowser#8278] [Show and tell]
Hello. I just made some scripts to interact with qutebrowser using rofi for menues and toml to declare search engines and quickmarks.

(Current) Features:

1. Uses toml for search engines and quickmarks, allowing to declare "templates" to generate uniform sets of aliases for websites like github (`gh.qb`, `ghi.qb`, `ghin.qb`, `ghp.qb`, `ghpn.qb`, etc. See [example](https://github.com/cortsf/quterofi?tab=readme-ov-file#example) in readme).
2. New ":open" menu allowing to switch the search engine of an unsubmitted search string, without having to move the cursor.
3. Allows to (quickly) switch search engines of an already submitted search string.
4. Allows to (quickly) edit already submitted search strings.
5. Allows to use the new `:open` menu from outside qb. It's possible to use it just to pick a search engine to search (in qb) the selected text inside a code editor or terminal, with minimal user input.

The menues are highly opinionated of course! One drawback of this approach is that it makes it a bit difficult to provide users with all the configuration options to tweak the rofi theme, key bindinds and everything else.

Performance is not great on slow computers. I've tried with a 400 lines toml file on a ryzen 5300U, and the open menu takes about 1/4 of a second to show. This could probably be [fixed/improved](https://github.com/cortsf/quterofi/issues/30). Although, I haven not verified if the opening of the rofi menu alone (without any parsing) is already laggy on slow computers.

https://github.com/cortsf/quterofi

@The-Compiler  do you think this new toml format could be made into qb, as an alternative opt-in experimental method for storing and/or reading quickmarks/search engines? I have been using another equivalent method for a long time, and it really helps to have such "uniform interfaces", specially for github repos.

```
c.use_experimental_toml_file = True
```

**Comment by The-Compiler** [comment]
> @The-Compiler do you think this new toml format could be made into qb, as an alternative opt-in experimental method for storing and/or reading quickmarks/search engines? I have been using another equivalent method for a long time, and it really helps to have such "uniform interfaces", specially for github repos.

From what I could gather, it seems to be the wrong approach to me. It's basically turning a declarative language into almost a programming language, that requires quite a lot of additional context to understand. I'd much rather have the ability to dynamically add quickmarks/bookmarks from `config.py` in a supported way (#7512 has a bit on it). That'd allow for the same kind of thing, but would IMHO be much easier to grasp, *and* offer way more flexibility too.

**Comment by cortsf** [comment]
Hm yes I personally don't care about stretching toml a bit. It was choosen because:

1. Json doesn't allows comments.
2. Parsers (and tooling like `toml2json`) exists everywhere.
3. This particular approach allows to define engines and quickmarks in a self-contained file that can be used from outside qb (the `quterofi/open` userscript can be used standalone..).
4. I never used yaml before, but was a bit familiar with toml.

I can't find too much info on #7512 (except for the link to #882). For me the important thing is to be able to read engines and quickmark definitions externally. If qb only supports a "plain" (no variables and such) json for engines, it's easy to generate this json externally , with toml, a form in a webpage or anything else. And it's also easy to read this json to call qb with url's from within code editors, terminals and so on.

**Comment by cortsf** [comment]
Just made a [bash script](https://github.com/cortsf/quterofi#dequterofi) to translate quterofi files into regular config.py engine declarations. This can be used to generate search engines dynamically without having to install quterofi. For quickmars, the `set_quickmarks` userscripts provides the same functionality (I guess it can be used standalone, without having to install it in qb).
